File: 31_oct_level_2/problem_18_elevator_control_system/solution.py
from typing import TextIO, List
import sys
from collections import deque

class Solution(object):
    def simulateElevatorSystem(self, numFloors, numElevators, requests):
        """
        :type numFloors: int
        :type numElevators: int
        :type requests: List[str]
        :rtype: List[str]
        """
        class Elevator:
            def __init__(self, elevator_id):
                self.id = elevator_id
                self.current_floor = 1
                self.direction = "IDLE"
                self.state = "IDLE"
                self.destinations = deque()
                self.busy = False
        
        elevators = [Elevator(i) for i in range(numElevators)]
        output = []
        step = 0
        
        for req in requests:
            parts = req.split()
            req_type = parts[0]
            
            if req_type == "EXTERNAL":
                floor = int(parts[1])
                direction = parts[2]
                
                # Simple dispatch: find nearest idle elevator
                best_elev = -1
                min_dist = float('inf')
                
                for i in range(numElevators):
                    if not elevators[i].busy:
                        dist = abs(elevators[i].current_floor - floor)
                        if dist < min_dist:
                            min_dist = dist
                            best_elev = i
                
                # If no idle elevator, use first one
                if best_elev == -1:
                    best_elev = 0
                
                elevators[best_elev].destinations.append(floor)
                elevators[best_elev].busy = True
                output.append(f"Step {step}: Elevator {best_elev} dispatched to floor {floor}")
                
            elif req_type == "INTERNAL":
                elev_id = int(parts[1])
                dest = int(parts[2])
                
                if 0 <= elev_id < numElevators:
                    elevators[elev_id].destinations.append(dest)
                    elevators[elev_id].busy = True
                    output.append(f"Step {step}: Elevator {elev_id} moving to floor {dest}")
            
            step += 1
        
        return output

def solve(infile: TextIO = sys.stdin, outfile: TextIO = sys.stdout) -> None:
    n, m, t = map(int, infile.readline().split())
    requests = []
    for _ in range(t):
        requests.append(infile.readline().strip())
    
    solution = Solution()
    output = solution.simulateElevatorSystem(n, m, requests)
    
    for line in output:
        outfile.write(line + '\n')

if __name__ == "__main__":
    solve()

File: 31_oct_level_2/problem_18_elevator_control_system/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <sstream>
#include <cmath>
#include <climits>
using namespace std;

class Solution {
private:
    enum Direction { IDLE, UP, DOWN };
    enum State { STOPPED, MOVING, DOORS_OPEN };
    
    struct Elevator {
        int id;
        int currentFloor;
        Direction direction;
        State state;
        queue<int> destinations;
        bool busy;
        
        Elevator(int id) : id(id), currentFloor(1), direction(IDLE), state(STOPPED), busy(false) {}
    };
    
public:
    vector<string> simulateElevatorSystem(int numFloors, int numElevators, 
                                          vector<string>& requests) {
        vector<Elevator> elevators;
        for (int i = 0; i < numElevators; i++) {
            elevators.push_back(Elevator(i));
        }
        
        vector<string> output;
        int step = 0;
        
        for (const string& req : requests) {
            istringstream iss(req);
            string type;
            iss >> type;
            
            if (type == "EXTERNAL") {
                int floor;
                string dir;
                iss >> floor >> dir;
                
                // Simple dispatch: find nearest idle elevator
                int bestElev = -1;
                int minDist = INT_MAX;
                
                for (int i = 0; i < numElevators; i++) {
                    if (!elevators[i].busy) {
                        int dist = abs(elevators[i].currentFloor - floor);
                        if (dist < minDist) {
                            minDist = dist;
                            bestElev = i;
                        }
                    }
                }
                
                // If no idle elevator, use first one
                if (bestElev == -1) {
                    bestElev = 0;
                }
                
                elevators[bestElev].destinations.push(floor);
                elevators[bestElev].busy = true;
                output.push_back("Step " + to_string(step) + 
                               ": Elevator " + to_string(bestElev) + 
                               " dispatched to floor " + to_string(floor));
            }
            else if (type == "INTERNAL") {
                int elevId, dest;
                iss >> elevId >> dest;
                
                if (elevId >= 0 && elevId < numElevators) {
                    elevators[elevId].destinations.push(dest);
                    elevators[elevId].busy = true;
                    output.push_back("Step " + to_string(step) + 
                                   ": Elevator " + to_string(elevId) + 
                                   " moving to floor " + to_string(dest));
                }
            }
            
            step++;
        }
        
        return output;
    }
};

void solve() {
    int n, m, t;
    cin >> n >> m >> t;
    cin.ignore();
    
    vector<string> requests;
    for (int i = 0; i < t; i++) {
        string line;
        getline(cin, line);
        requests.push_back(line);
    }
    
    Solution solution;
    vector<string> output = solution.simulateElevatorSystem(n, m, requests);
    
    for (const string& line : output) {
        cout << line << endl;
    }
}

int main() {
    solve();
    return 0;
}
